### DISCUSSION TITLE
thread 'main' panicked at '`spawn_local` called from outside of a `task::LocalSet`'

### DISCUSSION THREAD

**Discussion by afidegnum** [framesurge/perseus#187] [Q&A]
I have introduced the dynamic block using [transitions](https://github.com/afidegnum/sycatest/blob/master/src/main.rs)  into[ the component ](https://github.com/commune-org/linksapp/blob/main/src/components/index_hero.rs) and I'm facing the following perseus compilation error: 

```
warning: `linksapp` (bin "linksapp") generated 11 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 51.78s
     Running `dist/target_engine/debug/linksapp`
thread 'main' panicked at '`spawn_local` called from outside of a `task::LocalSet`', /home/afidegnum/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.1/src/task/local.rs:316:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

What could be the cause?

**Comment by arctic-hen7** [comment]
The file you've linked to has no Perseus imports, so I'm pretty sure this isn't a Perseus problem. The Perseus build process can't cause panics on its own, it will be your code that's somehow breaking. Beyond that, I'm not very familiar with Sycamore's suspense system, sorry, so I don't think I'll be much help on this one.

**Reply by afidegnum** [reply]
Can you please shed more light? I'm trying to import a sycamore's component into a perseus page which would respond to an event with a transition. 

What do you suggest to use?

**Reply by afidegnum** [reply]
from this [example](https://github.com/framesurge/perseus/blob/main/examples/demos/fetching/src/templates/index.rs), I noticed `perseus::spawn_local_scoped` is being used. How can it be used in this example?

**Reply by arctic-hen7** [reply]
I don't have time to review all your code right now, sorry. If you can create a minimal app that reproduces the problem, then I might be able to help further. Otherwise, I would recommend asking this question on the Sycamore repo. (Any Sycamore component should work fine with Perseus.)

**Reply by arctic-hen7** [reply]
Are you using this component inside a Perseus page? If you gave me the code for that, I might be of more use. There might be an issue with using async components, but there shouldn't be...


`spawn_local_scoped` is a Sycamore function that runs some async code in the browser. You can use it instead of an async component in your use case I think.

**Comment by afidegnum** [comment]
Here is the [code]( https://github.com/afidegnum/perseuspense). you can find it on[ gitpod ](https://afidegnum-perseuspense-hrxj0b5847c.ws-eu64.gitpod.io/)

**Reply by afidegnum** [reply]
I'm trying to implement suspense alternative in perseus using `spawn_local_scope` but I have few errors:
first, I am not able to get the default block before executing the function. 
secondly, `spawn_local_scope` is requesting `()` instead of a `View`.

```rs
spawn_local_scoped(cx, async move {
        TimeoutFuture::new(delay_ms).await;
        view! { cx,
            div {
                p { (if let Ok(content) = block.content(cx) { view! { cx, (content) } } else { view! { cx, "Oh no!" } }) }
            }
        }
    });
```

**Reply by arctic-hen7** [reply]
Have a `Signal` store your `View`, and then update it from the `spawn_local_scoped`.

**Reply by afidegnum** [reply]
ok,

**Reply by afidegnum** [reply]
Can you please check the updated repo? I managed to bring the view into `spawn_local_scoped` but I I'm greeted with an aditional error: 
```rs
error[E0277]: () doesn't implement std::fmt::Display
   --> src/templates/index.rs:110:5
    |
110 | /     view! { cx,
111 | |         div {
112 | |
...   |
134 | |         }
135 | |     }
    | |__^ () cannot be formatted with the default formatter
    |
```

**Reply by arctic-hen7** [reply]
I'll take a look later today.

**Reply by arctic-hen7** [reply]
Okay, rather than returning a `View` from `spawn_local_scoped`, you should instead create a `Signal` to store that `View`, and then use `.set()` on that `Signal` inside the `spawn_local_scoped`. Then use `.get()` on the `Signal` inside your proper `view!`, and it will automatically update from the `spawn_local_scoped`.

**Reply by afidegnum** [reply]
How do I also implement a transition effect and the suspense feature even if not entirely replicating sycamore's in spawn_local_scoped?

**Reply by arctic-hen7** [reply]
Sorry, I'm not familiar with how transitions work in Sycamore. @lukechu10 might be able to help?

**Reply by lukechu10** [reply]
It's not transitions that are causing the issue. It's the async that is the problem. I believe this is because somewhere in Perseus, the app is rendered outside of a `render_to_string` scope so there is no access to the `tokio::LocalSet` which should be provided by `provide_executor_scope`.

**Reply by arctic-hen7** [reply]
How does one render outside a `render_to_string` scope?

**Reply by lukechu10** [reply]
For example:
```rust
let node = App(cx);
let string = render_to_string(|_| node);
```
Here, `App` is being called outside the scope created by `render_to_string`.

**Reply by arctic-hen7** [reply]
Oh I see! I hadn't understood that. That shouldn't be too hard to fix, I'll have a look into it soon.

**Reply by arctic-hen7** [reply]
I've opened #225 for this, since there seems to be a bigger problem underlying this. Sorry it took me so long to get to! In the meantime, anyone wanting to use a `Suspense` in Perseus should only do so behind a `#[cfg(target_arch = "wasm32")]` target-gate.
